tests/test_serialization.py

```diff
@@ -0,0 +1,154 @@
+from typing import Optional, Type
+
+import attrs
+import pytest
+
+from web_poet import (
+    HttpResponse,
+    HttpResponseBody,
+    Injectable,
+    RequestUrl,
+    ResponseUrl,
+    WebPage,
+)
+from web_poet.page_inputs.url import _Url
+from web_poet.serialization import (
+    SerializedData,
+    deserialize,
+    read_serialized_data,
+    register_serialization,
+    serialize,
+    write_serialized_data,
+)
+
+
+def _compare_webpages(p1: WebPage, p2: WebPage) -> None:
+    assert p1.response.body == p2.response.body
+    assert p1.response.status == p2.response.status
+    assert p1.response.headers == p2.response.headers
+    assert p1.response._encoding == p2.response._encoding
+    _compare_urls(p1.response.url, p2.response.url)
+
+
+def _compare_urls(u1: _Url, u2: _Url) -> None:
+    assert type(u1) == type(u2)
+    assert str(u1) == str(u2)
+
+
+def test_serialization() -> None:
+    data = {"a": "b", "c": 42}
+    serialized_data = serialize(data)
+    deserialized_data = deserialize(dict, serialized_data)
+    assert data == deserialized_data
+
+
+def test_serialization_unsup() -> None:
+    class A:
+        pass
+
+    with pytest.raises(NotImplementedError):
+        serialize(A())
+
+    with pytest.raises(NotImplementedError):
+        deserialize(A, {})
+
+
+def test_serialization_webpage(book_list_html_response) -> None:
+    po = WebPage(book_list_html_response)
+    serialized_po = serialize(po)
+    deserialized_po = deserialize(WebPage, serialized_po)
+    _compare_webpages(po, deserialized_po)
+
+
+def test_serialization_httpresponse_encoding(book_list_html) -> None:
```

Comment:
The `encoding`/`_encoding` semantics were unclear to me so I wrote this. Looks like saving and restoring `_encoding` if it's not empty is correct.